datafusion/expr-common/src/groups_accumulator.rs

```diff
@@ -31,6 +31,13 @@ pub enum EmitTo {
     /// For example, if `n=10`, group_index `0, 1, ... 9` are emitted
     /// and group indexes '`10, 11, 12, ...` become `0, 1, 2, ...`.
     First(usize),
+    /// Emit all groups managed by blocks
+    AllBlocks,
+    /// Emit only the first `n` group blocks,
+    /// similar as `First`, but used in blocked `GroupValues` and `GroupAccumulator`.
+    ///
+    /// For example, `n=3`, `block size=4`, finally 12 groups will be returned.
+    FirstBlocks(usize),
```

Comment:
Rather than having two parallel emission modes for blocked output, I wonder if we could have some sort of "take" mode whose semantics did not shift the existing values down

For example, what if we introduced a notion of "block" across the group keys and and aggregators

```rust
pub enum EmitTo {
    /// Same
    All,
    /// Same
    First(usize),
    /// Takes the N'th block of rows from this accumulator
    /// it is an error to take the same batch twice or to emit `All` or `First` 
    /// after any TakeBatch(usize)
    TakeBlock(usize)
}
```

And then we would for example, make sure the group values and aggregators all saved data using blocks of 100K rows

Then to emit 1M rows, the accumulators would emit like
```
EmitTo::TakeBlock(0)
EmitTo::TakeBlock(1)
EmitTo::TakeBlock(2)
...
EmitTo::TakeBlock(9)
```

Reply:
It may be a great idea for reducing code changes(we dont need to refactor the returned emit value for T to Vec<T>).

Actually I found shifting the existing values down is only necessary for streaming aggr, because it need to record the current sort idx.

But we need to record the current block id in outer, it seems a bit complicated. Maybe we just define it like this?
We use the iterator approach to impl `AllBlocks` and `FirstBlocks` defined now?
If this makes sense, I can try switch the sketch to this way.
```
pub enum EmitTo {
    /// Same
    All,
    /// Same
    First(usize),
    /// Return the current block of rows from this accumulator
    /// We should always use this emit mode in blocked mode accumulator.
    CurrentBlock,
}
```

Reply:
> use the iterator approach to impl AllBlocks and FirstBlocks defined now

Not sure how does this work, but it looks like a neat idea. If we apply the same idea to "element" (First and All), and consider it as a specialized case with block_size = 1, I think we could end up a pretty nice abstraction. Probably we just need `EmitTo::Block(block_size)` 🤔  However, it is too far way from now. 😆

Reply:
> > use the iterator approach to impl AllBlocks and FirstBlocks defined now
> 
> Not sure how does this work, but it looks like a neat idea. If we apply the same idea to "element" (First and All), and consider it as a specialized case with block_size = 1, I think we could end up a pretty nice abstraction. Probably we just need `EmitTo::Block(block_size)` 🤔 However, it is too far way from now. 😆

🤔 Yes, other emit mode can indeed seen as a case with specialized blocke size in the iterator approach. But considered about performance, it is better to let `batch_size == block_size`.

After introduce the iterator approach, just 200+ codes to finished the  sketch, compared to the stale version sketch with 600+. 
The main work is just to add a stream state `ExecutionState::ProducingBlocks(blocks)` .
https://github.com/Rachelint/arrow-datafusion/blob/d79d912d1677549c825cafc405911973ace0df46/datafusion/physical-plan/src/aggregates/row_hash.rs#L728

Maybe it can show how the blocked optimzation works.